plonk/src/lagrange.rs

```diff
@@ -0,0 +1,224 @@
+//! Utilities for Lagrange interpolations, evaluations, coefficients for a
+//! polynomial
+
+use ark_ff::FftField;
+use ark_poly::{EvaluationDomain, Radix2EvaluationDomain};
+use ark_std::{ops::Range, vec, vec::Vec};
+
+// TODO: (alex) include these APIs upstream in arkworks directly
+
+/// A helper trait for computing Lagrange coefficients of an evaluation domain
+///
+/// from arkworks:
+/// Evaluate all Lagrange polynomials at tau to get the lagrange coefficients.
+/// Define the following as
+/// - H: The coset we are in, with generator g and offset h
+/// - n: The size of the coset H
+/// - Z_H: The vanishing polynomial for H. Z_H(x) = prod_{i in n} (x - hg^i) =
+///   x^n - h^n
+/// - v_i: A sequence of values, where v_0 = 1/(n * h^(n-1)), and v_{i + 1} = g
+///   * v_i
+///
+/// We then compute L_{i,H}(tau) as `L_{i,H}(tau) = Z_H(tau) * v_i / (tau - h *
+/// g^i)`
+#[allow(dead_code)]
+pub(crate) trait LagrangeCoeffs<F: FftField> {
+    /// Returns the first coefficient: `L_{0, Domain}(tau)`
+    fn first_lagrange_coeff(&self, tau: F) -> F;
+    /// Returns the last coefficient: `L_{n-1, Domain}(tau)`
+    fn last_lagrange_coeff(&self, tau: F) -> F;
+    /// Return a list of coefficients for `L_{range, Domain}(tau)`
+    fn lagrange_coeffs_for_range(&self, range: Range<usize>, tau: F) -> Vec<F>;
+}
+
+impl<F: FftField> LagrangeCoeffs<F> for Radix2EvaluationDomain<F> {
+    // L_0(tau) = Z_H(tau) * g^0 / (n * h^(n-1) *(tau - h * g^0))
+    // with g^0 = 1
+    // special care when tau in H, as both numerator and denominator is zero
+    fn first_lagrange_coeff(&self, tau: F) -> F {
+        let offset = self.coset_offset();
+        let z_h_at_tau = self.evaluate_vanishing_polynomial(tau);
+        if z_h_at_tau.is_zero() {
+            if tau == F::one() * offset {
+                F::one()
+            } else {
+                F::zero()
+            }
+        } else {
+            let denominator = self.size_as_field_element()
+                * offset.pow([self.size() as u64 - 1])
+                * (tau - offset);
+            z_h_at_tau * denominator.inverse().unwrap()
+        }
+    }
+
+    // L_n-1(tau) = Z_H(tau) * g^-1 / (n * h^(n-1) * (tau - h * g^-1))
+    // with g^n-1 = g^-1
+    fn last_lagrange_coeff(&self, tau: F) -> F {
```

Comment:
minor: there are many repetitive computations between `first_lagrange_coeff()` and `last_lagrange_coeff()`. When we need both coeffs, is it better to call `first_and_langrange_coeff()` or consider it as a special case of `lagrange_coeff_for_range()` API ? (Since {n, 1} is also a special range.)

Reply:
> consider it as a special case of lagrange_coeff_for_range() API ? (Since {n, 1} is also a special range.)

this is a little hard. 

but I agree that we could probably add a `first_and_last_langrange_coeff()` to save some repeated computation

Reply:
on a second thought, the only overlapped computation is `vanish_eval` (the `offset.pow()` is negligible, as we has offset=1), maybe we leave it as it is?

Reply:
added in [70ef5f5](https://github.com/EspressoSystems/jellyfish/pull/611/commits/70ef5f545728e2ef95abf2443abf002a1814f84d)

Reply:
Are we gonna using the implementation that saves repeated computation? If we're just calling first_lagrange_coeff() and last_lagrange_coeff() naively, maybe there is no need to add one more API.